### PR TITLE
feat: COSMOSHUB spec from MELLIFERA

### PR DESCRIPTION
**Pull request by RuslanGlaznyov**
**Name**: Ruslan  
**Discord**: ruslan_mellifera  
**Commission**: 3.5%  
- **+1%** - Support and run a Provider on testnet  
- **+0.5%** - Inherit Cosmos spec  
- **+1%** - Benchmark CU and timeouts (Need assistance from the team)  
- **+1%** - Provide active support to Providers, if more than 20 providers are retained over 3 months

---

My name is Ruslan, and I've been with Lava for more than 2 years. During this time, I have developed a range of services and integrations for LAVA, including:

- [Various Services](https://services.mellifera.network) (REST, RPC, snapshots, archive snapshots)
- SDK integration into the vote application ([1](https://github.com/mellifera-labs/janusbot?tab=readme-ov-file#lava-integration), [2](https://github.com/MELLIFERA-Labs/cosmolava-client))
- Ping Pub integration LAVA integration ([1](https://lava-explorer.mellifera.network))
- Support for Providers in both Testnet and Mainnet
- Support for RPC Consumer for testing
- And more...

I bring skills as both a programmer and a provider. If given the honor, I would gladly become a Cosmos Hub Spec Champion and support providers with all their needs. However, I acknowledge that I am not a highly active member of the Cosmos Hub community. If there are better candidates with a strong reputation within the Cosmos Hub community, I would gladly support them as a Spec Champion to foster great partnerships and development between Cosmos Hub and LAVA.

### CODE REVIEW

File: mainnet-1/specs/cosmoshub.json [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    "title": "Add Specs: Cosmos hub",
+    "description": "Adding new specification support for relaying cosmoshub data on Lava",
+    "specs": [
+      {
+        "index": "COSMOSHUB",
+        "name": "cosmos hub mainnet",
+        "enabled": true,
+        "imports": [
+          "COSMOSSDKFULL",
+          "COSMOSSDK45DEP"
```
Comment (omerlavanet) [inline mainnet-1/specs/cosmoshub.json:12]: this is outdated cosmos hub is already version 50

File: mainnet-1/specs/cosmoshub.json [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        "reliability_threshold": 268435455,
+        "data_reliability_enabled": true,
+        "block_distance_for_finalized_data": 0,
+        "blocks_in_finalization_proof": 1,
+        "average_block_time": 6500,
+        "allowed_block_lag_for_qos_sync": 2,
+        "shares": 1,
+        "min_stake_provider": {
+          "denom": "ulava",
+          "amount": "47500000000"
```
Comment (omerlavanet) [inline mainnet-1/specs/cosmoshub.json:23]: 47.5k LAVA is too high, let's do 15K LAVA as a standard for iprpc chains

File: mainnet-1/specs/cosmoshub.json [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  "proposal": {
+    "title": "Add Specs: Cosmos hub",
+    "description": "Adding new specification support for relaying cosmoshub data on Lava",
+    "specs": [
+      {
+        "index": "COSMOSHUB",
+        "name": "cosmos hub mainnet",
+        "enabled": true,
+        "imports": [
+          "COSMOSSDKFULL",
```
Comment (omerlavanet) [inline mainnet-1/specs/cosmoshub.json:11]: cosmossdk spec full is for v47, cosmos hub is already v50, we need to go over the apis, check what is missing and add them

File: mainnet-1/specs/cosmoshub.json [hunk 4]
```diff
@@ -0,0 +1,159 @@
+{
```
Comment (omerlavanet) [inline mainnet-1/specs/cosmoshub.json:1]: missing cosmos hub testnet spec, please add it too

File: mainnet-1/specs/cosmoshub.json [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+  "proposal": {
+    "title": "Add Specs: Cosmos hub",
+    "description": "Adding new specification support for relaying cosmoshub data on Lava",
+    "specs": [
+      {
+        "index": "COSMOSHUB",
+        "name": "cosmos hub mainnet",
+        "enabled": true,
+        "imports": [
```
Comment (omerlavanet) [inline mainnet-1/specs/cosmoshub.json:10]: when we import on mainnet we need to make sure all dependencies are also imported on mainnet and we don't have missing ones

File: mainnet-1/specs/cosmoshub.json [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+                  "block": 14200
+                }
+              }
+            ]
+          }
+        ],
+        "contributor": [
+          "lava@1r0n6tl824c05h9gu48ms88mcxgtea9v9et9xtc"
+        ],
+        "contributor_percentage": 0.1
```
Comment (omerlavanet) [inline mainnet-1/specs/cosmoshub.json:154]: the percentage should be 2.5%, the additional 1% for support and provider growth can only be added after 3 months

File: mainnet-1/specs/cosmoshub.json (hunk 2 above)
Comment (RuslanGlaznyov) [reply mainnet-1/specs/cosmoshub.json]: 👍

File: mainnet-1/specs/cosmoshub.json (hunk 4 above)
Comment (RuslanGlaznyov) [reply mainnet-1/specs/cosmoshub.json:1]: In mainnet we would have mainnet and testnet providers? I thought would be best to keep testnet in testnet and mainnet in mainnet

File: mainnet-1/specs/cosmoshub.json (hunk 1 above)
Comment (RuslanGlaznyov) [reply mainnet-1/specs/cosmoshub.json]: In cosmohub mainnet cosmossdk is v0.47 
In cosmosub testnet cosmossdk is v0.50  

We have only `COSMOSSDK` spec on the mainnet should we use only this or we need  something to add/change?

File: mainnet-1/specs/cosmoshub.json (hunk 6 above)
Comment (RuslanGlaznyov) [reply mainnet-1/specs/cosmoshub.json]: 👍

File: mainnet-1/specs/cosmoshub.json (hunk 4 above)
Comment (RuslanGlaznyov) [reply mainnet-1/specs/cosmoshub.json:1]: Do we plan to use archive extension for testnet spec?

Comment (omerlavanet) [reply mainnet-1/specs/cosmoshub.json:1]: @RuslanGlaznyov we will have both and monetize both with pools according to the rate the ecosystem wants, testnets need paid rpc too :)

**Comment by RuslanGlaznyov** [comment]
Added cosmossdk v50 spec: 

here is script I used for spec construction: 
```ts
import * as bun from "bun";

const specStarter = {
    "proposal": {
        "title": "Cosmos SDK v50 spec",
        "description": "Add cosmos sdk v50 spec",
        "specs": [
            {
                "index": "COSMOSSDK50",
                "enabled": false,
                "reliability_threshold": 268435455,
                "data_reliability_enabled": true,
                "block_distance_for_finalized_data": 0,
                "blocks_in_finalization_proof": 1,
                "average_block_time": "6500",
                "allowed_block_lag_for_qos_sync": "2",
                "block_last_updated": "910837",
                "min_stake_provider": {
                    "denom": "ulava",
                    "amount": "47500000000"
                },
                "providers_types": "dynamic",
                "imports": [
                    "IBC",
                    "TENDERMINT",
                    "COSMOSSDK"
                ],
                "api_collections": [
                    {
                        "enabled": true,
                        "collection_data": {
                            "api_interface": "rest",
                            "internal_path": "",
                            "type": "GET",
                            "add_on": ""
                        },
                        "apis": [],
                        "headers": [],
                        "parse_directives": [],
                        "verifications": [],
                        "extensions": []
                    },
                    {
                        "enabled": true,
                        "collection_data": {
                            "api_interface": "grpc",
                            "internal_path": "",
                            "type": "",
                            "add_on": ""
                        },
                        "apis": [],
                        "headers": [],
                        "parse_directives": [],
                        "verifications": [],
                        "extensions": []
                    },
                ]

            }
        ]
    },
    "deposit": "7000000000ulava"
}

function defaultRestApiConstruct(api: string) {
    return {
        "enabled": true,
        "name": api,
        "compute_units": "10",
        "extra_compute_units": "0",
        "category": {
            "deterministic": true,
            "local": false,
            "subscription": false,
            "stateful": 0,
            "hanging_api": false
        },
        "block_parsing": {
            "parser_arg": [
                "latest"
            ],
            "parser_func": "DEFAULT",
            "default_value": "",
            "encoding": ""
        },
        "timeout_ms": "0"
    }
}
function defaultGrpcApiConstruct(api: string) {
    return {
        "enabled": true,
        "name": api,
        "compute_units": "10",
        "extra_compute_units": "0",
        "category": {
            "deterministic": true,
            "local": false,
            "subscription": false,
            "stateful": 0,
            "hanging_api": false
        },
        "block_parsing": {
            "parser_arg": [
                "latest"
            ],
            "parser_func": "DEFAULT",
            "default_value": "",
            "encoding": ""
        },
        "timeout_ms": "0"
    }
}
function disableConstruct(api: string) {
    return {
        "enabled": false,
        "name": api,
    }
}

// have api in cosmos node and don't have in COSMOSSDK spec
const apiToADDCosmos = [
    "/cosmos/bank/v1beta1/denoms_metadata_by_query_string", "/cosmos/bank/v1beta1/denom_owners_by_query",
    "/cosmos/base/node/v1beta1/status", "/cosmos/distribution/v1beta1/{owner_address}/tokenize_share_record_rewards",
    "/cosmos/gov/v1/constitution", "/cosmos/staking/v1beta1/tokenize_share_records", "/cosmos/staking/v1beta1/last_tokenize_share_record_id",
    "/cosmos/staking/v1beta1/tokenize_share_lock_info/{address}", "/cosmos/staking/v1beta1/tokenize_share_record_by_denom/{denom}",
    "/cosmos/staking/v1beta1/tokenize_share_record_by_id/{id}", "/cosmos/staking/v1beta1/tokenize_share_record_owned/{owner}",
    "/cosmos/staking/v1beta1/total_liquid_staked", "/cosmos/staking/v1beta1/total_tokenize_shared_assets"
]
// have grpc in cosmos node and don't have in COSMOSSDK spec
const grpcToADDCosmos = [
    "cosmos.autocli.v1.Query/AppOptions", "cosmos.bank.v1beta1.Query/DenomMetadataByQueryString",
    "cosmos.bank.v1beta1.Query/DenomOwnersByQuery", "cosmos.base.node.v1beta1.Service/Status",
    "cosmos.consensus.v1.Query/Params", "cosmos.distribution.v1beta1.Query/TokenizeShareRecordReward",
    "cosmos.gov.v1.Query/Constitution", "cosmos.reflection.v1.ReflectionService/FileDescriptors",
    "cosmos.staking.v1beta1.Query/AllTokenizeShareRecords", "cosmos.staking.v1beta1.Query/LastTokenizeShareRecordId",
    "cosmos.staking.v1beta1.Query/TokenizeShareLockInfo", "cosmos.staking.v1beta1.Query/TokenizeShareRecordByDenom",
    "cosmos.staking.v1beta1.Query/TokenizeShareRecordById", "cosmos.staking.v1beta1.Query/TokenizeShareRecordsOwned",
    "cosmos.staking.v1beta1.Query/TotalLiquidStaked", "cosmos.staking.v1beta1.Query/TotalTokenizeSharedAssets"
]

// have in COSMOSSDK spec and don't have in cosmos node
const apiToDisableCosmos = [
    "/cosmos/bank/v1beta1/balances/{address}/{denom}", "/cosmos/bank/v1beta1/base_denom",
    "/cosmos/bank/v1beta1/supply/{denom}", "/cosmos/bank/v1beta1/supply_without_offset",
    "/cosmos/bank/v1beta1/supply_without_offset/{denom}", "/cosmos/evidence/v1beta1/evidence/{evidence_hash}",
    "/cosmos/group/v1/vote_by_proposal_voter/{proposal_id}/{voter}", "/cosmos/group/v1/proposals/{proposal_id}/tally",
    "/cosmos/group/v1/group_policies_by_group/{group_id}", "/cosmos/group/v1/proposals_by_group_policy/{address}",
    "/cosmos/group/v1/votes_by_proposal/{proposal_id}", "/cosmos/group/v1/group_policies_by_admin/{admin}"
]

// have in COSMOSSDK spec and don't have in cosmos node
const grpcToDisableCosmos = [
    "cosmos.bank.v1beta1.Query/BaseDenom", "cosmos.bank.v1beta1.Query/SupplyOfWithoutOffset",
    "cosmos.bank.v1beta1.Query/TotalSupplyWithoutOffset", "cosmos.group.v1.Query/GroupsByAdmin",
    "cosmos.group.v1.Query/GroupPolicyInfo", "cosmos.group.v1.Query/GroupInfo", "cosmos.group.v1.Query/VoteByProposalVoter",
    "cosmos.group.v1.Query/TallyResult", "cosmos.group.v1.Query/ProposalsByGroupPolicy",
    "cosmos.group.v1.Query/VotesByVoter", "cosmos.group.v1.Query/VotesByProposal", "cosmos.group.v1.Query/GroupsByMember",
    "cosmos.group.v1.Query/Proposal", "cosmos.group.v1.Query/Groups", "cosmos.group.v1.Query/GroupMembers",
    "cosmos.group.v1.Query/GroupPoliciesByGroup", "cosmos.group.v1.Query/GroupPoliciesByAdmin"
]
// have api in cosmos node and don't have in COSMOSSDK spec
const apiToAddIBC = [
    "/ibc/apps/fee/v1/channels/{channel_id}/relayers/{relayer}/counterparty_payee", "/ibc/apps/fee/v1/channels/{channel_id}/ports/{port_id}/fee_enabled",
    "/ibc/apps/fee/v1/fee_enabled", "/ibc/apps/fee/v1/channels/{packet_id.channel_id}/ports/{packet_id.port_id}/sequences/{packet_id.sequence}/incentivized_packet",
    "/ibc/apps/fee/v1/incentivized_packets", "/ibc/apps/fee/v1/channels/{channel_id}/ports/{port_id}/incentivized_packets",
    "/ibc/apps/fee/v1/channels/{channel_id}/relayers/{relayer}/payee", "/ibc/apps/fee/v1/channels/{packet_id.channel_id}/ports/{packet_id.port_id}/sequences/{packet_id.sequence}/total_ack_fees",
    "/ibc/apps/fee/v1/channels/{packet_id.channel_id}/ports/{packet_id.port_id}/sequences/{packet_id.sequence}/total_recv_fees",
    "/ibc/apps/fee/v1/channels/{packet_id.channel_id}/ports/{packet_id.port_id}/sequences/{packet_id.sequence}/total_timeout_fees",
    "/ibc/apps/transfer/v1/denom_traces/{hash=**}", "/ibc/core/channel/v1/params", "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/upgrade",
    "/ibc/core/channel/v1/channels/{channel_id}/ports/{port_id}/upgrade_error", "/ibc/apps/packetforward/v1/params"
]
// have grpc in cosmos node and don't have in COSMOSSDK spec
const grpcToAddIBC = [
    "ibc.applications.fee.v1.Query/CounterpartyPayee", "ibc.applications.fee.v1.Query/FeeEnabledChannel",
    "ibc.applications.fee.v1.Query/FeeEnabledChannels", "ibc.applications.fee.v1.Query/IncentivizedPacket",
    "ibc.applications.fee.v1.Query/IncentivizedPackets", "ibc.applications.fee.v1.Query/IncentivizedPacketsForChannel",
    "ibc.applications.fee.v1.Query/Payee", "ibc.applications.fee.v1.Query/TotalAckFees",
    "ibc.applications.fee.v1.Query/TotalRecvFees", "ibc.applications.fee.v1.Query/TotalTimeoutFees",
    "ibc.core.channel.v1.Query/ChannelParams", "ibc.core.channel.v1.Query/NextSequenceSend",
    "ibc.core.channel.v1.Query/Upgrade", "ibc.core.channel.v1.Query/UpgradeError", "ibc.core.client.v1.Query/VerifyMembership"
]
// have in COSMOSSDK spec and don't have in cosmos node
const apiToDisableIBC =[
    "/ibc/apps/router/v1/params",
    "/ibc/apps/transfer/v1/denom_hashes/{trace}",
    "/ibc/apps/transfer/v1/denom_traces/{hash}",
    "/ibc/client/v1/params"
];
// have in COSMOSSDK spec and don't have in cosmos node
const grpcToDisableIBC: string[]  = [];


// rest apis
const restApi = specStarter.proposal.specs[0].api_collections[0].apis
const allAddRest = [...apiToADDCosmos, ...apiToAddIBC]
const allDisableRest = [...apiToDisableCosmos, ...apiToDisableIBC]
for (const api of allAddRest) {
    restApi.push(defaultRestApiConstruct(api))
}
for (const api of allDisableRest) {
    restApi.push(disableConstruct(api))
}
// grpc apis
const grpcApi = specStarter.proposal.specs[0].api_collections[1].apis
const allAddGrpc = [...grpcToADDCosmos, ...grpcToAddIBC]
const allDisableGrpc = [...grpcToDisableCosmos, ...grpcToDisableIBC]
for (const api of allAddGrpc) {
    grpcApi.push(defaultGrpcApiConstruct(api))
}
for (const api of allDisableGrpc) {
    grpcApi.push(disableConstruct(api))
}

await bun.write('COSMOSSDK50.json', JSON.stringify(specStarter, null, 2)) 
```

**Comment by RuslanGlaznyov** [comment]
Added cosmwasm spec. this almost same spec from cookbook but added few apis: 
Rest: "/cosmwasm/wasm/v1/contract/build_address"
Grpc:  [ "cosmwasm.wasm.v1.Query/BuildAddress", "cosmwasm.wasm.v1.Query/ContractsByCreator"]

**Comment by RuslanGlaznyov** [comment]
Pushed cosmsohub spec. COSMOSHUBT and COSMOSHUB are the same because they have same binary version

File: mainnet-1/specs/cosmossdkv50.json [hunk 7]
```diff
@@ -0,0 +1,1465 @@
...
+                    "latest"
+                  ],
+                  "parser_func": "DEFAULT",
+                  "default_value": "",
+                  "encoding": ""
+                }
+              },
+              {
+                "enabled": false,
+                "name": "/cosmos/bank/v1beta1/balances/{address}/{denom}",
```
Comment (omerlavanet) [inline mainnet-1/specs/cosmossdkv50.json:625]: parser func default latest, and spec cagetory deterministic are needed in all of the next ones

File: mainnet-1/specs/cosmossdkv50.json [hunk 8]
```diff
@@ -0,0 +1,1465 @@
...
+                    "latest"
+                  ],
+                  "parser_func": "DEFAULT",
+                  "default_value": "",
+                  "encoding": ""
+                }
+              },
+              {
+                "enabled": false,
+                "name": "cosmos.bank.v1beta1.Query/BaseDenom",
```
Comment (omerlavanet) [inline mainnet-1/specs/cosmossdkv50.json:1371]: parser func default latest and category deterministic missing here too

**Review by omerlavanet — CHANGES_REQUESTED** [review]

**Review by omerlavanet — APPROVED** [review]

**Comment by omerlavanet** [comment]
* ruslan has added missing apis for cosmos version 50, and added missing apis for cosmos hub such as feemarket and interchain_security
* ruslan has ran the specs using unitests for sanity
* i have verified integrity of the spec inlcuding the contributor of 3% (creating the spec, supporting and running a provider on testnet, benchmarking cus and timeouts)
* cus and timeouts benchmarking will be added in the coming month (delivery date: by the 25th Sep) as an additional PR

**Pull request merged**